File: src/test/java/org/meetkt/catalogue/BarcodeScannerUnitTest.java
package org.meetkt.catalogue;

import org.junit.jupiter.api.Test;
import org.meetkt.catalogue.domain.model.Catalogue;
import org.meetkt.catalogue.domain.model.Product;
import org.meetkt.catalogue.domain.model.ProductId;
import org.meetkt.catalogue.exception.NoProductFoundForBarcodeException;

import java.util.Collections;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.meetkt.catalogue.fixture.ProductFixture.aProduct;

class BarcodeScannerUnitTest {

    @Test
    void shouldThrowAnExceptionGivenProductDoesNotExistForTheBarcode() {
        String barcode = "no-product-found-for-this-barcode";
        Catalogue catalogue = new Catalogue(Collections.emptyList());

        BarcodeScanner barcodeScanner = new BarcodeScanner(catalogue);

        assertThrows(NoProductFoundForBarcodeException.class,
                () -> barcodeScanner.scanOrThrow(barcode)
        );
    }

    @Test
    void shouldReturnAProductForABarcode() {
        String barcode = "product-001-barcode";
        Catalogue catalogue = new Catalogue(List.of(aProduct().withProductId("P-001").withBarcode(barcode).build()));
        BarcodeScanner barcodeScanner = new BarcodeScanner(catalogue);

        Product product = barcodeScanner.scanOrThrow(barcode);

        assertThat(product.id()).isEqualTo(new ProductId("P-001"));
    }
}

File: src/test/kotlin/in/meetkt/catalogue/BarcodeScannerUnitTest.kt
package `in`.meetkt.catalogue

import `in`.meetkt.catalogue.domain.model.Catalogue
import `in`.meetkt.catalogue.domain.model.ProductId
import `in`.meetkt.catalogue.fixture.ProductFixture
import io.mockk.every
import io.mockk.mockk
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test

class BarcodeScannerUnitTest {

    @Test
    fun shouldReturnAnExceptionGivenProductDoesNotExistForTheBarcode() {
        val barcode = "no-product-found-for-this-barcode"
        val catalogue = mockk<Catalogue>()
        every { catalogue.productWith(barcode) } returns null
        val barcodeScanner = BarcodeScanner(catalogue)

        val product = barcodeScanner.scanOrThrow(barcode)

        assertThat(product.isLeft()).isTrue
    }

    @Test
    fun shouldReturnAProductForABarcode() {
        val barcode = "product-001-barcode"
        val catalogue = mockk<Catalogue>()
        val barcodeScanner = BarcodeScanner(catalogue)
        every { catalogue.productWith(barcode) } returns
                ProductFixture.aProduct().withProductId("001").withBarcode(barcode).build()

        val product = barcodeScanner.scanOrThrow(barcode)

        assertThat(product.orNull()?.id).isEqualTo(ProductId("001"))
    }
}
